ibm/service/iamidentity/resource_ibm_iam_trusted_profile_template_assignment.go

```diff
@@ -0,0 +1,766 @@
+	}
+	if err = d.Set("last_modified_by_id", templateAssignmentResponse.LastModifiedByID); err != nil {
+		return diag.FromErr(fmt.Errorf("error setting last_modified_by_id: %s", err))
+	}
+	if err = d.Set("entity_tag", templateAssignmentResponse.EntityTag); err != nil {
+		return diag.FromErr(fmt.Errorf("error setting entity_tag: %s", err))
+	}
+
+	return nil
+}
+
+func resourceIBMTrustedProfileTemplateAssignmentUpdate(context context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	iamIdentityClient, err := meta.(conns.ClientSession).IAMIdentityV1API()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	updateTrustedProfileAssignmentOptions := &iamidentityv1.UpdateTrustedProfileAssignmentOptions{}
+	updateTrustedProfileAssignmentOptions.SetAssignmentID(d.Id())
+	updateTrustedProfileAssignmentOptions.SetIfMatch(d.Get("entity_tag").(string))
+
+	hasChange := false
+
+	if d.HasChange("template_version") {
+		updateTrustedProfileAssignmentOptions.SetTemplateVersion(int64(d.Get("template_version").(int)))
+		hasChange = true
+	}
+
+	if hasChange || d.Get("status") == "failed" { // allow the same version to retry failed assignments
+		_, response, err := iamIdentityClient.UpdateTrustedProfileAssignmentWithContext(context, updateTrustedProfileAssignmentOptions)
+		if err != nil {
+			log.Printf("[DEBUG] UpdateTrustedProfileAssignmentWithContext failed %s\n%s", err, response)
+			return diag.FromErr(fmt.Errorf("UpdateTrustedProfileAssignmentWithContext failed %s\n%s", err, response))
+		}
+
+		_, err = waitForAssignment(meta, d, isTrustedProfileTemplateAssigned)
+		if err != nil {
+			return diag.FromErr(fmt.Errorf("error assigning %s", err))
+		}
+	}
+
+	return resourceIBMTrustedProfileTemplateAssignmentRead(context, d, meta)
+}
+
+func resourceIBMTrustedProfileTemplateAssignmentDelete(context context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	iamIdentityClient, err := meta.(conns.ClientSession).IAMIdentityV1API()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	deleteTrustedProfileAssignmentOptions := &iamidentityv1.DeleteTrustedProfileAssignmentOptions{}
+
+	deleteTrustedProfileAssignmentOptions.SetAssignmentID(d.Id())
+
+	_, response, err := iamIdentityClient.DeleteTrustedProfileAssignmentWithContext(context, deleteTrustedProfileAssignmentOptions)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("[DEBUG] DeleteTrustedProfileAssignmentWithContext failed %s\n%s", err, response))
+	}
+
+	_, err = waitForAssignment(meta, d, isTrustedProfileAssignmentRemoved)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error removing assignment %s", err))
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func waitForAssignment(meta interface{}, d *schema.ResourceData, refreshFn func(string, interface{}) resource.StateRefreshFunc) (interface{}, error) {
+	log.Printf("Waiting for  (%s) to complete...", d.Id())
+
+	timeoutStr, isSet := os.LookupEnv(envTimeoutDurationKey)
```

Comment:
Use Terraform timeouts declared in resource schema. 
https://developer.hashicorp.com/terraform/plugin/sdkv2/resources/retries-and-customizable-timeouts

envTimeoutDurationKey is not valid way to read timeouts.

Reply:
Done.